libs/core/src/bloom_filter.cpp

```diff
@@ -0,0 +1,375 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/bitvector.hpp"
+#include "core/bloom_filter.hpp"
+#include "core/byte_array/const_byte_array.hpp"
+#include "crypto/fnv.hpp"
+
+#include <openssl/evp.h>
+
+#include <cassert>
+#include <cstddef>
+#include <functional>
+#include <vector>
+
+// Aim for 1 false positive per this many positive queries
+constexpr std::size_t const INVERSE_TARGET_FALSE_POSITIVE_RATE = 100000u;
+
+// No point in evaluating the filter's quality until this many positive queries
+// have been executed
+constexpr std::size_t const MEANINGFUL_STATS_THRESHOLD = 5u * INVERSE_TARGET_FALSE_POSITIVE_RATE;
+
+constexpr std::size_t const INITIAL_SIZE_IN_BITS = 8 * 10 * 1024 * 1024;
+
+namespace fetch {
+namespace internal {
+
+HashSourceFactory::HashSourceFactory(HashSourceFactory::Functions hash_functions)
```

Comment:
Can we add documentation to all of these methods please.

Reply:
Done